bitcoin/src/witness_program.rs

```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
+            self.version
+        };
+        output.push(encoded_version);
         output.push(self.program.len() as u8);
-        output.copy_from_slice(&self.program);
+        output.extend_from_slice(&self.program);
         output
     }
-
-    /// A BIP173 version conversion utility function.
-    /// Convert a given version to a value between 0 and 16 or OP_1 through OP_16.
-    pub fn convert_version(version: u8) -> u8 {
-        if version > 0x00 && version <= 0x10 {
-            // encode OP_1 through OP_16
-            version + 0x50
-        } else if version > 0x50 {
-            // decode OP_1 through OP_16
-            version - 0x50
-        } else {
-            // OP_0
-            version
-        }
-    }
 }
 
 impl FromStr for WitnessProgram {
-    type Err = &'static str;
+    type Err = WitnessProgramError;
 
     /// Returns a witness program given its hex representation.
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         WitnessProgram::new(&match hex::decode(s) {
             Ok(bytes) => bytes,
-            Err(_) => return Err("Error decoding hex string"),
+            Err(_) => return Err(WitnessProgramError::ProgramDecodingError),
         })
     }
 }
 
-impl fmt::Display for WitnessProgram {
-    /// Prints a witness program in hex representation.
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", hex::encode(self.to_vec()))
+#[cfg(test)]
+mod tests {
+    mod test_from_str {
+        use std::str::FromStr;
+
+        use crate::witness_program::{WitnessProgram, WitnessProgramError};
+
+        const VALID_P2SH_P2WPKH_PROGRAM: &str = "0014751e76e8199196d454941c45d1b3a323f1433bd6";
+        const INVALID_P2SH_P2WPKH_PROGRAM_LENGTH: &str = "0014751e76e8199196d454941c45d1b3a323f143";
+
+        #[test]
+        fn from_str_invalid_p2sh_p2wpkh_program_len() {
+            match WitnessProgram::from_str(INVALID_P2SH_P2WPKH_PROGRAM_LENGTH) {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
+                Err(e) => {
+                    assert_eq!(WitnessProgramError::MismatchedProgramLength(18, 20), e)
+                }
+            }
+        }
+
+        #[test]
+        fn from_str_valid_p2sh_p2wpkh_program() {
+            match WitnessProgram::from_str(VALID_P2SH_P2WPKH_PROGRAM) {
+                Ok(witness_program) => {
+                    assert_eq!(0x00, witness_program.version);
+                    assert_eq!(vec![0x75, 0x1e, 0x76, 0xe8, 0x19,
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
+            }
+        }
+
+        const VALID_P2SH_P2WSH_PROGRAM: &str = "00201863143c14c5166804bd19203356da136c985678cd4d27a1b8c6329604903262";
```

Comment:
Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682